flow360/component/utils.py

```diff
@@ -553,3 +562,107 @@ def storage_size_formatter(size_in_bytes):
     if size_in_bytes < 1024**3:
         return f"{size_in_bytes / (1024 ** 2):.2f} MB"
     return f"{size_in_bytes / (1024 ** 3):.2f} GB"
+
+
+# pylint: disable=too-few-public-methods
+class HasId(Protocol):
+    """
+    Protocol for objects that have an `id` attribute.
+
+    Attributes
+    ----------
+    id : str
+        Unique identifier for the asset.
+    """
+
+    id: str
+
+
+AssetT = TypeVar("AssetT", bound=HasId)
+
+
+class ProjectAssetCache(Generic[AssetT]):
+    """
+    A cache to manage project assets with a unique ID system.
+
+    Attributes
+    ----------
+    current_id : str, optional
+        The ID of the currently set asset.
+    cache : dict of str to AssetT
+        Dictionary storing assets with their IDs as keys.
+    """
+
+    current_id: str = None
+    cache: dict[str, AssetT] = {}
+
+    def get_asset(self, asset_id: str = None) -> AssetT:
```

Comment:
Interesting that the API behaves this way. I discussed the extent of functionality of the Project class with Maciej before and we agreed that `Project.from_cloud` should only retrieve the root asset in this iteration, but I agree that it would be best if we could retrieve all assets here. I'll add it in a separate PR.